Santosh Kumar Gajawada: Update Readme regarding the pub/sub, test cas… (#17)
WeirdMagician/conduit-connector-google-cloudstorage: README.md

# Conduit Connector Google Cloud Storage

### General
The Google Cloud Storage connector is one of [Conduit](https://github.com/ConduitIO/conduit) plugins. It provides source GCS connectors.

### How to build it
Run `make`.

### Testing

Run `make test` with optional `GOTEST_FLAGS` set to run all the tests. You must set the environment variables (`GCP_ServiceAccount_Key`
, `GCP_ProjectID`, `GCP_Bucket`)
before you run all the tests. If not set, the tests that use these variables will be ignored.

Cases/Scenarios which are dependent on GCS Error response and where these can't be reproducible are ignored in the test cases.
 

> Integration test cases are written to cover all of the scenarios of snapshot and cdc iterators except when tomb dead as it happens only in the above mentioned case.

Reviewer: I don't think we should mention tomb in the readme.. a bit too technical

Author: ok

 
## GCS Source

> The Google Cloud Storage Source Connector connects to a GCS bucket with the provided configurations, using serviceAccountKey(`JSON`) and bucket(`BucketName`) details . Then will call `Configure` to parse the

Reviewer: JSON here is confusing.. it's a format, it's not the value nor the key itself
```suggestion
The Google Cloud Storage Source Connector connects to a GCS bucket with the provided configurations, using `serviceAccountKey` and `bucket` details from the configurations. Then will call `Configure` to parse the
```

Author: Done

* To capture "delete" actions, the GCS bucket versioning must be enabled.
* To capture "insert" or "update" actions, the bucket versioning doesn't matter.

> If the object as "N" versions in a bucket only the live version is considered as a Record.

Reviewer: ```suggestion
If the object has multiple versions in a bucket, then only the live version is considered a Record.
```

The connector goes through two reading modes.

* Snapshot mode (Value 0): which loops through the GCS bucket and returns the objects that are already in there. The _position type_ during this mode is 0. which makes the connector know at what mode it is and what object it last
 read. The _position Timestamp_ will be used when changing to CDC mode, the iterator will capture changes that
 happened after that.

* CDC mode: (Value 1) this mode iterates through the GCS bucket every `pollingPeriod` and captures new actions made on the bucket.
 the _Position Type_ during this mode is 1. This position is used to return only the
 actions with a _Position Timestamp_ higher than the last record returned even if the timestamp got matched then the decision would be based on lexicographical comparison, which will ensure that no duplications are in
 place.

>  The CDC mode will start after a single action(Insert/Update/Delete) completed post snapshot.

Reviewer: does it start CDC after the first change? or after the last record from snapshot? we should have an integration test for (if the connector stops after snapshot and starts again.. it should start from CDC)

Author: GCS object iterator doesn't support the HasNext function to check whether the there is a next object. So in the current implementation CDC will start after the first change.

Let me know explore are there any options to support "if the connector stops after snapshot and starts again.. it should start from CDC"

Author: @maha-hajja  There is no direct method provide from the GCS API to know the last item. But i went through the internal implementation 
when PageInfo.Token is empty and PageInfo.Remaining() is zero then that is the last item of the snapshot iterator. so where we can switch it back to CDC.
I think will track this issue separately. Pls share your observations

Reference Doc:
https://pkg.go.dev/google.golang.org/api/iterator#PageInfo
https://pkg.go.dev/google.golang.org/api/iterator#PageInfo.Remaining

Author: Issue #23

| `bucket`          | the GCS bucket name                                                                 | yes       | "bucket_name"       |
| `pollingPeriod`       | polling period for the CDC mode, formatted as a time.Duration string. default is "1s"  | no        | "2s", "500ms"       |

> When testing from the swagger or any rest client provide the stringified value in `serviceAccountKey`

Reviewer: ```suggestion
When testing using swagger or any rest client, provide the JSON stringified value in `serviceAccountKey`
```

Author: Done

> Example: console.log(JSON.stringify(JSON.stringify(key file content))) run in the JavaScript to get the preferred stringified value because wrapping JSON around the single quotes is not accepted in swagger OR

Reviewer: ```suggestion
Example: `console.log(JSON.stringify(JSON.stringify(key file content)))` run this JavaScript command to get the preferred stringified value. because wrapping JSON with single quotes is not accepted in swagger OR
```

Author: Thanks its Done
